In [2]:
import numpy as np
import pandas as pd

In [6]:
phone1_df = pd.read_csv(r'rough_data\phone_user_reviews_kaggle\phone_user_review_file_1.csv')
phone2_df = pd.read_csv(r'rough_data\phone_user_reviews_kaggle\phone_user_review_file_2.csv')
phone3_df = pd.read_csv(r'rough_data\phone_user_reviews_kaggle\phone_user_review_file_3.csv')
phone4_df = pd.read_csv(r'rough_data\phone_user_reviews_kaggle\phone_user_review_file_4.csv')
phone5_df = pd.read_csv(r'rough_data\phone_user_reviews_kaggle\phone_user_review_file_5.csv')
phone6_df = pd.read_csv(r'rough_data\phone_user_reviews_kaggle\phone_user_review_file_6.csv')

In [9]:
phone1_df.head(10)

phone_url       date lang country            source  \
0  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us  Verizon Wireless   
1  /cellphones/samsung-galaxy-s8/  4/28/2017   en      us       Phone Arena   
2  /cellphones/samsung-galaxy-s8/   5/4/2017   en      us            Amazon   
3  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us           Samsung   
4  /cellphones/samsung-galaxy-s8/  5/11/2017   en      us  Verizon Wireless   
5  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
6  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
7  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
8  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   
9  /cellphones/samsung-galaxy-s8/  5/10/2017   en      us  Verizon Wireless   

                domain  score  score_max  \
0  verizonwireless.com   10.0       10.0   
1       phonearena.com   10.0       10.0   
2           amazon.com    6.0       10.0   
3          samsung.com    9.2       10.0   
4  verizonwireless.com    4.0       10.0   
5  verizonwireless.com   10.0       10.0   
6  verizonwireless.com    4.0       10.0   
7  verizonwireless.com    6.0       10.0   
8  verizonwireless.com    6.0       10.0   
9  verizonwireless.com    8.0       10.0   

                                             extract           author  \
0  As a diehard Samsung fan who has had every Sam...       CarolAnn35   
1  Love the phone. the phone is sleek and smooth ...        james0923   
2  Adequate feel. Nice heft. Processor's still sl...         R. Craig   
3  Never disappointed. One of the reasons I've be...      Buster2020    
4  I've now found that i'm in a group of people t...       S Ate Mine   
5  I am the type of person who never would comple...            BDB76   
6  The way this Samsung S8 phone operates is more...         KLC30306   
7  I bought this phone very excited to use it. I ...             Mnhy   
8  It is an extremely advanced and truly a Smart ...          Mr Alan   
9  Great phone with a PHENOMENAL Camera, not all ...  Raven Wolf Song   

                                             product  
0                                  Samsung Galaxy S8  
1                                  Samsung Galaxy S8  
2  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                      Samsung Galaxy S8 64GB (AT&T)  
4                                  Samsung Galaxy S8  
5                                  Samsung Galaxy S8  
6                                  Samsung Galaxy S8  
7                                  Samsung Galaxy S8  
8                                  Samsung Galaxy S8  
9                                  Samsung Galaxy S8

In [10]:
phone2_df.head(10)

phone_url        date lang country  source        domain  \
0  /cellphones/leagoo-lead-7/   4/15/2015   en      us  Amazon    amazon.com   
1  /cellphones/leagoo-lead-7/   5/23/2015   en      gb  Amazon  amazon.co.uk   
2  /cellphones/leagoo-lead-7/   4/27/2015   en      gb  Amazon  amazon.co.uk   
3  /cellphones/leagoo-lead-7/   4/22/2015   en      gb  Amazon  amazon.co.uk   
4  /cellphones/leagoo-lead-7/   4/18/2015   en      gb  Amazon  amazon.co.uk   
5  /cellphones/leagoo-lead-7/   4/14/2015   en      us  Amazon    amazon.com   
6  /cellphones/leagoo-lead-7/    4/1/2015   en      gb  Amazon  amazon.co.uk   
7  /cellphones/leagoo-lead-7/   11/6/2015   it      it  Amazon     amazon.it   
8  /cellphones/leagoo-lead-7/  10/26/2015   it      it  Amazon     amazon.it   
9  /cellphones/leagoo-lead-7/   9/19/2015   it      it  Amazon     amazon.it   

   score  score_max                                            extract  \
0    2.0       10.0  The telephone headset is of poor quality , not...   
1   10.0       10.0  This is my first smartphone so I have nothing ...   
2    8.0       10.0  Great phone. Battery life not great but seems ...   
3   10.0       10.0      Best 90 quid I've ever spent on a smart phone   
4   10.0       10.0  I m happy with this phone.it s very good.thx team   
5   10.0       10.0  Exelent phone. Evriting perfect. Service is am...   
6   10.0       10.0  Great mobile, but contains malware, in the for...   
7   10.0       10.0  maneggevole, pratico, veloce, ottime foto........   
8   10.0       10.0  Prodotto ECCELLENTE!!! Praticamente perfetto, ...   
9   10.0       10.0  Cercavo un telefono che avesse una fotocamera ...   

          author                                            product  
0           luis  Leagoo Lead7 5.0 Inch HD JDI LTPS Screen 3G Sm...  
1     Mark Lavin  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
2         tracey  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
3  Reuben Ingram  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
4         viorel  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
5         Milena  Leagoo Lead7 5.0 Inch HD JDI LTPS Screen 3G Sm...  
6    lee carroll  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
7     Filippo L.  LEAGOO Lead 7 MTK6582 Quad Core 5" Pollice Sma...  
8    Emiliano S.  LEAGOO Lead 7 MTK6582 Quad Core 5" Pollice Sma...  
9        rossano  LEAGOO Lead 7 MTK6582 Quad Core 5" Pollice Sma...

In [11]:
phone3_df.head(10)

phone_url       date lang country  \
0  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  11/7/2015   pt      br   
1  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  10/2/2015   pt      br   
2  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/2/2015   pt      br   
3  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/2/2015   pt      br   
4  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/1/2015   pt      br   
5  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/1/2015   pt      br   
6  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   8/3/2015   pt      br   
7  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  7/29/2015   pt      br   
8  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  7/24/2015   pt      br   
9  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  7/21/2015   pt      br   

      source            domain  score  score_max  \
0  Submarino  submarino.com.br    6.0       10.0   
1  Submarino  submarino.com.br   10.0       10.0   
2  Submarino  submarino.com.br   10.0       10.0   
3  Submarino  submarino.com.br    8.0       10.0   
4    Colombo    colombo.com.br    8.0       10.0   
5  Submarino  submarino.com.br    8.0       10.0   
6  Submarino  submarino.com.br   10.0       10.0   
7  Submarino  submarino.com.br    4.0       10.0   
8  Submarino  submarino.com.br   10.0       10.0   
9  Submarino  submarino.com.br   10.0       10.0   

                                             extract  \
0  recomendo, eu comprei um, a um ano, e agora co...   
1  Comprei um pouco desconfiada do site e do celu...   
2  Muito bom o produto, obvio que tem versões mel...   
3  Unica ressalva fica para a camera que poderia ...   
4  Rapidez e atenção na entrega. O aparelho é mui...   
5  Gostei do produto e entrega ocorreu antes do p...   
6  ótimo aparelho, unica coisa ruim é a camera fr...   
7  Aparelho muito simples pelo preço que se paga,...   
8  excelente produto gostei muito, estou satifeit...   
9  È um excelente aparelho,leve prático ....câmer...   

                           author  \
0                herlington tesch   
1             Luisa Silva Marieta   
2                           Cyrus   
3       Marcela Santa Clara Brito   
4  Claudine Maria Kuhn Walendorff   
5                            JCVS   
6                   Juliana Viana   
7                     Julio cabal   
8         RIVALDO CIOQUETA JUNIOR   
9                         Dulcili   

                                             product  
0  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
1  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
2  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
3  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
4  Smartphone Samsung Galaxy S3 Slim, Dual Chip, ...  
5  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
6  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
7  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
8  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
9  Samsung Smartphone Samsung Galaxy S3 Slim G381...

In [12]:
phone4_df.head(10)

phone_url        date lang country  \
0  /cellphones/samsung-s7262-duos-galaxy-ace/   3/11/2015   en      us   
1  /cellphones/samsung-s7262-duos-galaxy-ace/  17/11/2015   en      in   
2  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
3  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
4  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
5  /cellphones/samsung-s7262-duos-galaxy-ace/  27/10/2015   en      in   
6  /cellphones/samsung-s7262-duos-galaxy-ace/  27/10/2015   en      us   
7  /cellphones/samsung-s7262-duos-galaxy-ace/  26/10/2015   en      in   
8  /cellphones/samsung-s7262-duos-galaxy-ace/  25/10/2015   en      in   
9  /cellphones/samsung-s7262-duos-galaxy-ace/  21/10/2015   en      in   

   source      domain  score  score_max  \
0  Amazon  amazon.com    2.0       10.0   
1  Zopper  zopper.com   10.0       10.0   
2  Amazon   amazon.in    4.0       10.0   
3  Amazon   amazon.in    6.0       10.0   
4  Amazon   amazon.in   10.0       10.0   
5  Amazon   amazon.in    8.0       10.0   
6  Amazon  amazon.com    2.0       10.0   
7  Amazon   amazon.in    2.0       10.0   
8  Amazon   amazon.in    6.0       10.0   
9  Amazon   amazon.in   10.0       10.0   

                                             extract            author  \
0  was not conpatable with my phone as stated. I ...  Frances DeSimone   
1  Decent Functions and Easy to Operate Pros:- Th...     Expert Review   
2  Not Good Phone such price. Hang too much and v...   Amazon Customer   
3                               not bad for features   Amazon Customer   
4                                  Excellent product               NHK   
5                           Good in reasonable price            Rupali   
6  I hate it no flash no front camera microphone ...            Ashney   
7  The phone is getting struck very frequently.. ...   Amazon Customer   
8  Original product received with complete packag...   Amazon Customer   
9  Good product from Samsung. Just right for ever...   Amazon Customer   

                                             product  
0  Samsung Galaxy Star Pro DUOS S7262 Unlocked Ce...  
1                Samsung Galaxy Star Pro S7262 Black  
2  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
3  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
4  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
5  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
6  Samsung Galaxy Star Pro DUOS S7262 Unlocked Ce...  
7  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
8  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
9  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)

In [13]:
phone5_df.head(10)

phone_url        date lang country      source  \
0  /cellphones/karbonn-k1616/   7/13/2016   en      in  91 Mobiles   
1  /cellphones/karbonn-k1616/   7/13/2016   en      in  91 Mobiles   
2  /cellphones/karbonn-k1616/   7/13/2016   en      in  91 Mobiles   
3  /cellphones/karbonn-k1616/   4/25/2014   en      in     Naaptol   
4  /cellphones/karbonn-k1616/   4/23/2013   en      in     Naaptol   
5  /cellphones/karbonn-k1616/   2/24/2013   en      in     Naaptol   
6  /cellphones/karbonn-k1616/   2/23/2013   en      in     Naaptol   
7  /cellphones/karbonn-k1616/    2/7/2013   en      in     Naaptol   
8  /cellphones/karbonn-k1616/   1/12/2013   en      in     Naaptol   
9  /cellphones/karbonn-k1616/  10/13/2012   en      in     Naaptol   

          domain  score  score_max  \
0  91mobiles.com    2.0       10.0   
1  91mobiles.com    6.0       10.0   
2  91mobiles.com    4.0       10.0   
3    naaptol.com   10.0       10.0   
4    naaptol.com   10.0       10.0   
5    naaptol.com    8.0       10.0   
6    naaptol.com   10.0       10.0   
7    naaptol.com    8.0       10.0   
8    naaptol.com    8.0       10.0   
9    naaptol.com    8.0       10.0   

                                             extract             author  \
0  I bought 1 month before. currently speaker is ...          venkatesh   
1  I just bought one week back, I have Airtel con...             Venkat   
2  one problem in this handset opera is not worki...             krrish   
3  here Karbonn comes up with an another excellen...    BRIJESH CHAUHAN   
4  What a phone, all so on Naaptol my god 23% off...      Suraj CHAUHAN   
5  here Karbonn comes up with an another excellen...      Prasoon Gupta   
6  the camera quality is good compared even with ...    Yogeet Patnayak   
7  this mobile having basic features with compare...       Murali Avula   
8  Karbonn mobiles are well known for there quali...  Jeseel Cholassery   
9  The phone is handy, good build quality, howeve...      Sobhit tiwari   

                 product  
0          Karbonn K1616  
1          Karbonn K1616  
2          Karbonn K1616  
3  Karbonn K1616 - Black  
4  Karbonn K1616 - Black  
5  Karbonn K1616 - Black  
6  Karbonn K1616 - Black  
7  Karbonn K1616 - Black  
8  Karbonn K1616 - Black  
9  Karbonn K1616 - Black

In [14]:
phone6_df.head(10)

phone_url        date lang country  \
0  /cellphones/samsung-instinct-sph-m800/   9/16/2011   en      us   
1  /cellphones/samsung-instinct-sph-m800/   2/13/2014   en      us   
2  /cellphones/samsung-instinct-sph-m800/  12/30/2011   en      us   
3  /cellphones/samsung-instinct-sph-m800/  10/18/2008   en      us   
4  /cellphones/samsung-instinct-sph-m800/    9/6/2008   en      us   
5  /cellphones/samsung-instinct-sph-m800/   1/28/2009   en      us   
6  /cellphones/samsung-instinct-sph-m800/   7/25/2008   en      us   
7  /cellphones/samsung-instinct-sph-m800/   5/12/2015   en      us   
8  /cellphones/samsung-instinct-sph-m800/   4/26/2015   en      us   
9  /cellphones/samsung-instinct-sph-m800/   9/11/2014   en      us   

                      source             domain  score  score_max  \
0                Phone Arena     phonearena.com    8.0       10.0   
1                     Amazon         amazon.com    6.0       10.0   
2                Phone Scoop     phonescoop.com    9.0       10.0   
3              HandCellPhone  handcellphone.com    4.0       10.0   
4               Reviewed.com       reviewed.com    6.0       10.0   
5               ReviewStream   reviewstream.com    NaN        NaN   
6  Personal Electronics buzz         pebuzz.com    NaN        NaN   
7                     Amazon         amazon.com    4.0       10.0   
8                     Amazon         amazon.com    6.0       10.0   
9                     Amazon         amazon.com    4.0       10.0   

                                             extract           author  \
0  I've had the phone for awhile and it's a prett...       ajabrams95   
1  to be clear it is not the sellers fault that t...        Stephanie   
2  Well i love this phone. i have had ton of phon...         snickers   
3  I have had my Instinct for several months now ...              A4C   
4  i have had this instinct phone for about two m...        betaBgood   
5  When I heard that my friend had gotten the new...             rich   
6  This is a great multi-media phone. The screen ...        Highlight   
7  For the price I expected the phone not to be s...          jessica   
8          Good for a temporary phone for a preteen.  Amazon Customer   
9          Didn't work. Helpful in getting a refund.  Tamarra Kemsley   

                     product  
0        Samsung Instinct HD  
1  Samsung SPH M800 Instinct  
2              Instinct M800  
3           Samsung Instinct  
4           Samsung Instinct  
5     Samsung Instinct Phone  
6           Samsung Instinct  
7  Samsung SPH M800 Instinct  
8  Samsung SPH M800 Instinct  
9  Samsung SPH M800 Instinct

In [15]:
combined_phone_df = pd.concat([phone1_df, phone2_df, phone3_df, phone4_df, phone5_df, phone6_df], axis=0, ignore_index=True)

In [18]:
combined_phone_df

phone_url       date lang country  \
0           /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
1           /cellphones/samsung-galaxy-s8/  4/28/2017   en      us   
2           /cellphones/samsung-galaxy-s8/   5/4/2017   en      us   
3           /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
4           /cellphones/samsung-galaxy-s8/  5/11/2017   en      us   
...                                    ...        ...  ...     ...   
1415128  /cellphones/alcatel-ot-club_1187/  5/12/2000   de      de   
1415129  /cellphones/alcatel-ot-club_1187/  5/11/2000   de      de   
1415130  /cellphones/alcatel-ot-club_1187/   5/4/2000   de      de   
1415131  /cellphones/alcatel-ot-club_1187/   5/1/2000   de      de   
1415132  /cellphones/alcatel-ot-club_1187/  4/25/2000   de      de   

                   source               domain  score  score_max  \
0        Verizon Wireless  verizonwireless.com   10.0       10.0   
1             Phone Arena       phonearena.com   10.0       10.0   
2                  Amazon           amazon.com    6.0       10.0   
3                 Samsung          samsung.com    9.2       10.0   
4        Verizon Wireless  verizonwireless.com    4.0       10.0   
...                   ...                  ...    ...        ...   
1415128              Ciao              ciao.de    2.0       10.0   
1415129              Ciao              ciao.de   10.0       10.0   
1415130              Ciao              ciao.de    2.0       10.0   
1415131              Ciao              ciao.de    8.0       10.0   
1415132              Ciao              ciao.de    2.0       10.0   

                                                   extract        author  \
0        As a diehard Samsung fan who has had every Sam...    CarolAnn35   
1        Love the phone. the phone is sleek and smooth ...     james0923   
2        Adequate feel. Nice heft. Processor's still sl...      R. Craig   
3        Never disappointed. One of the reasons I've be...   Buster2020    
4        I've now found that i'm in a group of people t...    S Ate Mine   
...                                                    ...           ...   
1415128  Weil mein Onkel bei ALcatel arbeitet habe ich ...    david.paul   
1415129  Hy Liebe Leserinnen und Leser!! Ich habe seit ...  Christiane14   
1415130  Jetzt hat wohl Alcatell gedacht ,sie machen wa...    michaelawr   
1415131  Ich bin seit 2 Jahren (stolzer) Besitzer eines...   claudia0815   
1415132  Was sich Alkatel hier wieder ausgedacht hat,sc...    michaelawr   

                                                   product  
0                                        Samsung Galaxy S8  
1                                        Samsung Galaxy S8  
2        Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                            Samsung Galaxy S8 64GB (AT&T)  
4                                        Samsung Galaxy S8  
...                                                    ...  
1415128                            Alcatel Club Plus Handy  
1415129                            Alcatel Club Plus Handy  
1415130                            Alcatel Club Plus Handy  
1415131                            Alcatel Club Plus Handy  
1415132                            Alcatel Club Plus Handy  

[1415133 rows x 11 columns]

In [47]:
combined_phone_df['phone_url'].value_counts()[:1885]

/cellphones/samsung-galaxy-s-iii/      17093
/cellphones/apple-iphone-5s/           16379
/cellphones/samsung-galaxy-s6/         16145
/cellphones/samsung-galaxy-s5/         16082
/cellphones/samsung-galaxy-s7-edge/    15917
                                       ...  
/cellphones/samsung-sgh-b130/            100
/cellphones/alcatel-ot-735/              100
/cellphones/pantech-pocket/               99
/cellphones/sony-ericsson-v600i/          99
/cellphones/zte-nubia-z5s-mini/           99
Name: phone_url, Length: 1885, dtype: int64

1,882 Unique phones have 25+ reviews (before filtering out amazon)

In [49]:
combined_phone_df['phone_url'].value_counts()[:2657]

/cellphones/samsung-galaxy-s-iii/      17093
/cellphones/apple-iphone-5s/           16379
/cellphones/samsung-galaxy-s6/         16145
/cellphones/samsung-galaxy-s5/         16082
/cellphones/samsung-galaxy-s7-edge/    15917
                                       ...  
/cellphones/xiaomi-mi4s/                  50
/cellphones/alcatel-ot-525/               50
/cellphones/nokia-7270/                   50
/cellphones/doro-phoneeasy-740/           50
/cellphones/lg-ls855/                     49
Name: phone_url, Length: 2657, dtype: int64

2,656 Unique phones have 50+ reviews (before filtering out amazon)

In [50]:
combined_phone_df['phone_url'].value_counts()[:3487]

/cellphones/samsung-galaxy-s-iii/      17093
/cellphones/apple-iphone-5s/           16379
/cellphones/samsung-galaxy-s6/         16145
/cellphones/samsung-galaxy-s5/         16082
/cellphones/samsung-galaxy-s7-edge/    15917
                                       ...  
/cellphones/o2-ice/                       25
/cellphones/micromax-andro-a60/           25
/cellphones/samsung-sgh-x650/             25
/cellphones/lg-kp260/                     24
/cellphones/inq-cloud-touch/              24
Name: phone_url, Length: 3487, dtype: int64

3,486 Unique phones have 25+ reviews (before filtering out amazon)

In [53]:
combined_phone_df['score_max'].value_counts()

10.0    1351644
Name: score_max, dtype: int64

In [54]:
combined_phone_df['score'].value_counts()

10.0    656239
8.0     296018
2.0     128485
6.0     116616
4.0      72462
         ...  
0.4          2
8.9          1
6.1          1
4.9          1
1.5          1
Name: score, Length: 86, dtype: int64

All ratings are out of 10

In [55]:
combined_phone_df['score'].mean()

8.007059551183373

In [57]:
combined_phone_df['source'].value_counts()

Amazon             728471
Yandex             123066
Ciao                59425
Samsung             45585
MercadoLibre        33531
                    ...  
RECENZJE.NET.PL         1
 \tStaples              1
Mighty Ape              1
Foto.no                 1
Shop Your Way           1
Name: source, Length: 331, dtype: int64

331 different sources of reviews, with about half of all reviews coming from amazon.

In [56]:
# Commented out this line to avoid accidental overwriting of the file
# combined_phone_df.to_csv(r'rough_data\phone_user_reviews_kaggle\combined_phone_reviews.csv')

Next step is to get rid of the unneeded columns, and to filter out data from amazon (already have it represented in my project).